### DISCUSSION TITLE
POLL: Should `time` restrict the version of `serde` that is used?

### DISCUSSION THREAD

**Discussion by jhpratt** [time-rs/time#611] [Seeking input]
Context: https://github.com/serde-rs/serde/issues/2538

There is a roughly 8 second improvement in compile time for the initial (non-incremental) build, with incremental builds being slightly slower. My thoughts are present [here](https://github.com/serde-rs/serde/issues/2538#issuecomment-1684517372). One correction is that it is not a 3 millisecond improvement, but a 3 millisecond _regression_ per invocation.

Due to the change in `serde` being a potential security vulnerability, I changed the semver requirement to be `<= 1.0.171`, which excludes all versions with pre-built binaries. This has caused a fair amount of disagreement about the reasonableness of my action.

## Voting

If you **support** the change and want versions of `serde` with pre-built binaries to be unsupported by `time`, please react with :+1:.

If you **do not support** the change and want all versions of `serde` to be supported by `time`, including those with pre-built binaries, please react with :-1:

All other reactions will be ignored.

This poll will remain open until [2022-08-22 0:00 UTC](https://www.tickcounter.com/countdown/4387589). After voting closes, a new release will be made if necessary. For the result to be valid, a minimum of 20 total votes are required.

## Discussion

I will be leaving the discussion unlocked, but will not be responding to any messages. However, all replies will be monitored for compliance with the code of conduct.

---

## Results

The results can be viewed [here](https://github.com/time-rs/time/discussions/611#discussioncomment-6786275).

**Comment by epage** [comment]
Something missing from this is the implications of this change.  By putting an upper bound on a version requirement, it can make impossible to resolve dependency combinations that are confusing to end users, especially when these are transitive dependencies (e.g. [reddit thread](https://www.reddit.com/r/rust/comments/p8clcx/how_to_fix_cargo_dependency_issue/))

See https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#multiple-requirements for more details

**Reply by pinkforest** [reply]
Probably long run only.

IMO on short term `time` can patch-bump and relax it when this has been vetted and ironed out more at ecosystem lvl

fwiw - It could have been two release streams where major 1 gives you from source and major 2 would give the binary

This would have been informed choice IMO opting into more experimental builds with binaries

The hard truth now is there will be fragmentation regardless - whether it's the top level binary or libraries in the middle

**Comment by wezm** [comment]
I think the pinning also prevents swapping serde with something like https://github.com/commons-rs/serde-deblobbed since the latter uses the newer version number (see note in their README).

**Reply by pinkforest** [reply]
That serde-deblobbed patch is broken regardless as it doesn't provide alternative versions and .patch doesn't operate that way.

This just will fragment the ecosystem there is no question of that and nothing will change here given either pin-relaxed or not.

**Reply by decathorpe** [reply]
I plan to publish a serde-compatible (if possible) fork on crates.io under a different name, I just didn't have the time to do that yet. The -deblobbed fork was just a first step that I could do relatively quickly, but it's only a short-term / stop-gap solution.

Also, what do you mean "patch doesn't operate that way"? I tested the change locally and it looked like it worked (except for a version conflict due to the `time` crate that I noted in the README).

**Comment by AlexTMjugador** [comment]
First of all, I appreciate that this poll is being done and that there is a plan in place to deal with whatever decision `serde` makes going forward (see the context linked in the first post). I think both are necessary to handle this situation well.

- - -

However, I'd like to continue discussing the security implications of this change, as I believe they are more nuanced than they might appear at first glance, and those nuances may justify changing the actions taken. I also think that, in general, `time` is not in a good position to make such decisions on behalf of its users. The following paragraphs will be a follow up to what I said in https://github.com/time-rs/time/commit/500f8e4517cb89f0aa80087130ed2f8b8349f162#r124957986, where I discussed the possibility that crates can already run arbitrary code on build via build scripts and procedural macros.

> That code can be reviewed (and frequently is by organizations). The binary cannot.

That's a reasonable _a priori_ view, but it's not complete. To begin the discussion, let's keep in mind what I think we can all agree are the essential items to review when assessing whether or not a software supply chain is secure:

- The origin of the supply chain (i.e., where do you get binaries, source code and other build artifacts from). This determines the set of vulnerabilities a malicious actor can use to compromise your security objectives.
- The content of the supply chain (i.e., what binaries and build artifacts your build system ends up with). This determines if the supply chain does bad things.

In the case of the precompiled `serde_derive` binary, its [origin is the `serde_derive` crate itself](https://docs.rs/crate/serde_derive/1.0.183/source/), as far as I can tell. That binary targets Linux systems, so the exposure of any threat in it is limited to Linux systems. _That reduces or even nullifies its risk in cases where Windows, MacOS or BSD systems are used for builds._ (Which means that, as things stand, this measure does nothing but potentially create unsolvable dependency trees for Windows, MacOS and BSD users.) In turn, the `serde_derive` crate is either downloaded from `crates.io`, or read from a local vendor copy, for folks that use, for example, `cargo vendor`.

The relevant contents of `serde_derive` for this security discussion are the source code of its build scripts, procedural macros, and this precompiled Linux binary. Therefore, a supply chain security policy must define:

- The process to decide how to trust `serde_derive`'s source code.
- The process to decide how to trust `serde_derive`'s precompiled binary.

Approaches to deciding whether to trust `serde_derive`'s source include (but are not limited to) performing code reviews, or trusting the maintainers and other parts of the ecosystem to "do the right thing". (Indeed, if the opportunity and time costs of enforcing a stricter process are greater than the value and/or cost of recovering the information assets you are dealing with, the optimal choice is to "be lazy".)

On the other hand, the approaches to deciding whether to trust `serde_derive`'s precompiled binary are fundamentally similar. The difference is that it is _usually_ harder to verify that a binary is safe to use, especially when it's not clear how exactly it was built. But note the emphasis on the word "usually": source code can also be obfuscated, at which point it is no longer directly readable by humans, and should be treated like an opaque binary. (By the way, given how smart David seems to be, judging by the code he writes, I think that, if he was up to something nasty, he would agree that it is far more stealthy and effective to obfuscate some malicious payload on the build scripts or proc macros that most people trust without further ado than distribute a precompiled binary that will make security-conscious people wary.)

But the possibility of source code being obfuscated is not the only reason why I said "usually". As much as I don't like it, the vast majority of computers in the world are running binaries whose source code and build process are not publicly known: from Intel Management Engine firmware to NVIDIA drivers, passing through proprietary operating systems and beyond, there are many binaries with unknown source code running on a typical system, and any one of them can create security vulnerabilities. In practice, how do people even get some sense sense of security with all these black boxes of code? There are several approaches, and some of them can be easier than reviewing source code under some circumstances:

- Do not any run binary whose build process is not known for certain. This is the trivial solution to the problem, but it may have prohibitively high usability costs, as there may be no practical alternatives for them. In the context of the `time` crate, this means not running the `serde_derive` precompiled binary. This can be achieved by:
  - Using mandatory access control checks (e.g., AppArmor, SELinux) to prevent `serde_derive` from spawning processes. ([But this would make it panic](https://github.com/serde-rs/serde/blob/bfcd44704f847ac5a9f3072e102e803b5ebbef31/precompiled/serde_derive/src/lib_precompiled.rs#L35-L55), which in my opinion is not a good way to handle an error from which one can recover from by falling back to traditional proc macro expansion.)
  - Targeting `musl` instead of `glibc` for Linux platforms, or CPU architectures other than x64: currently, [`serde_derive` only uses the precompiled binary for the `x86_64-unknown-linux-gnu` target](https://github.com/serde-rs/serde/blob/bfcd44704f847ac5a9f3072e102e803b5ebbef31/precompiled/serde_derive/src/lib.rs#L18-L22).
- Analyze the behavior of a particular binary in a sandbox environment, and/or use static analysis tools such as disassemblers to reverse engineer its logic (edit: to be fair, reverse engineering optimized machine code is indeed pretty difficult, but note that this is only one of several possible approaches), and keep track of changes made to that binary. If you can be sure that a binary is safe to use no matter how it was built, and that binary hasn't been modified, then you can be sure that binary does not introduce vulnerabilities. In the context of the `serde_derive` crate, this means verifying that the precompiled binaries are secure, and extending that trust to an approved set of binaries. This can be done by comparing hashes of the `serde_derive` crate files from a cryptographically secure hash function.
- Sandboxing the build process. If builds are always done in isolated environments (e.g., well-secured containers, dedicated virtual machines with limited network access), malicious code may not even have a chance to compromise valuable information assets.

Given this walkthrough of how would a responsible and resourceful developer would evaluate security vulnerabilities and ways to address them, I think it's abundantly clear that the solution of enforcing stricter than necessary version requirements on `serde` is ill-suited to tackle security concerns. It's reasonable to define threat models where the cost of breaking a build is much greater than the cost (perhaps already paid) of enforcing other security policies with respect to precompiled binaries, and not only `time` can't know that, but it forces a vulnerability prevention choice that may be inappropriate for a given policy. (It is no coincidence that none of the ways I suggested to address security concerns with these binaries deal with semver requirements.)

As I said in other posts, the proper way for `time` to deal with this security issue is to at least contact `serde_derive` developers about it, and get them to commit to an action plan. With an action plan in place, security risks can be better assessed, and more informed decisions can be made faster, with hopefully less collateral damage to people who rely on the "ecosystem doing the right thing for them" approach.

By the way, may I remember you all of the fact fact that, as far as I know, three weeks have gone by without `serde_derive` effectively addressing these prebuilt binaries concerns? That's plenty of time for a library such as `serde` to try to compromise lots of information assets, if it wanted to do so. If we are really serious about security, we should focus our resources on getting `serde_derive` to do things better, not working around its changes on individual crates, which is an inefficient use of developer time.

**Comment by pinkforest** [comment]
This is basic supply-chain security ... and a lot about trust when it's lost

Given this change was never done in a manner that ensured

 0) .. **reasonable disclosure / informed opt-in** of this fact

 1) .. **reproducible builds** were possible / straightforward

 2) .. **auditability** and fitness into existing build workflows beforehand breaking them

 3) .. **development** environment now running non-auditable/reproducible binaries

 4) .. **opt-out** mechanism that would mitigate above

This exposes potentially the `time` development environment to unaudited binaries creating uncertainty at the development stage

Even if someone for once makes auditable / reproducible build it will take long time until this gets into the tooling given wide ecosystem.

This change flew through unnoticed - granted https://crates.io/crates/cackle gave a heads-up 

It was easy to be naive when it was not properly documented though.

Many FOSS things require informed "non-free" or so opt-in - not by default - mechanism to binary blobs

Using precompiled binaries at libraries is IMO not the way to go and I am glad time made the choice of not opting into the bins.

This isn't just a security issue but this is a build issue that was supposed to be "Experiment" given changelog.

Does that tell anyone about opting into something knowingly ? I don't think so.

Besides that it takes agency away and this was done silently in the supply-chain.

We should not be operating at reverse-engineering binaries level or changing my workflows to audit any upstream code changes.

The code-changes as per current workflows are designed for source-level changes and not running binary-level reproducible builds.

Sorry but until the system evolves - these kind of changes should not be pushed like this and anyone is welcome to opt-out.

And by opting out,

`time` will save their contributors and their environments from exposure of the non-audited changing binary dependencies.

Let's keep in mind that people have different build & development environments.

And the fact that it is reversible and temporary which can be relaxed after this has been vetted and ironed out.

So needless to say I am woting :+1:

**Reply by AlexTMjugador** [reply]
I think this is a fairly reasonable security assessment, but it should be directed towards `serde_derive` maintainers, not `time`. [As I thoroughly explained](https://github.com/time-rs/time/discussions/611#discussioncomment-6772753), `time` can only do so much to mitigate exposure and be aware of the wider security policy, so it is in an ill-posed position to enforce security decisions. "Saving [...] contributors and their environments from exposure of the non-audited changing binary dependencies" must be balanced against causing breakage and churn that could be better addressed by contacting `serde_derive` directly. (By the way, the "binary dependencies" may not be changing at all in lots of projects due to the usage of `Cargo.lock` files, which individual maintainers can choose to keep pinned to older `serde` versions if they want.)

Edit:

> I am glad time made the choice of not opting into the bins.

As it stands, this is more of an imposition to end users of `time` than a choice. `serde_derive` does not have built-in ways to opt out of those binaries in the applicable environments, and I think that's the fundamental issue we should try to address instead.

**Reply by pinkforest** [reply]
I am working both ends ICYMI - I'm trying hard to get opt-out mechanism at least

Cargo.lock gets overwritten all the time. This is about protecting the development environments.

`time` can choose to support whatever `serde` version it chooses.

And `time` not supporting given serde version is implementation detail at best especially given no SemVer breakage.

There is nothing controversial about what `time` is doing and less so than opting into binaries at project development.

**Reply by AlexTMjugador** [reply]
> `time` can choose to support whatever `serde` version it chooses.
> 
> And `time` not supporting given serde version is implementation detail at best especially given no SemVer breakage.
> 
> There is nothing controsersial about what `time` is doing and less so than opting into binaries at project development.

As Ed said, not supporting some `serde` versions is no longer an implementation detail when it causes unsolvable dependency trees. It also can stop being an implementation detail when it causes several versions of `serde` to be built, which negatively impacts build times.

My point is that `time` is not in a good position to take these decisions, and so it shouldn't. Security vulnerabilities should be fixed at their source, so they have the widest fix possible, and not worked around like this. (Edit: or at least be better thought out; if my analysis is correct, non-Linux platforms are not even affected by this.)

**Reply by pinkforest** [reply]
Possibly in the long run but hopefully when that becomes a bottleneck we will have that opt-out I'm working currently.

There is going to be fragmenation regardless - `time` can patch-bump and relax it when this has been vetted and ironed out.

**Reply by AlexTMjugador** [reply]
Of course, some fragmentation is to be expected now, so that's fair. For what it's worth, I'm looking forward to that opt-out too, for reasons that go beyond security.

**Reply by danakj** [reply]
This is well stated. Some ecosystem fracture around serde is now required unless/until they (or Rust) provide a way to avoid the binary distribution. By avoiding the prebuilts in other libraries it helps avoid further fracture around them as well.

If time ever required serde with prebuilts, then time would also no longer be a viable dependency either for projects that care about supply chain security, auditability and reproducibility. Or complex and unreliable local patching would be needed but that doesn’t feel like a long term nor even really a good short term solution. The prebuilts are unfortunately viral.

**Comment by pacak** [comment]
1. serde is not the only crate that can use faster compiling proc macros and I'm afraid once other maintainers realize that they also can get away with it - dealing with blob macros can turn into an exciting game of whack-a-mole.
2. compilation time for derive macro is not even a problem, compilation time for the generated code is. in prod on my $day_job switching to precompiled blob (even if it worked - it does not) will make exactly zero speedup since serde derive macro is not in a critical path, in fact it will make it slower due to binary spawn overhead. But even then, shaving off 10 seconds of about 20 minutes of the compilation wall-time that can be attributed to serde directly... :tada:?
3. this problem should probably addressed by cargo with wasm.

Decisions by both `serde` and `time` are wrong from the rust ecosystem point of view, but one by `time` moves things back where they supposed to be so I support them. :+1:

**Comment by AlexTMjugador** [comment]
I'd like to mention that @pinkforest [has submitted a PR to `serde` addressing these security concerns](https://github.com/serde-rs/serde/pull/2580). I think that some readers of this poll may find it interesting to take a look at it.

**Comment by andreisilviudragnea** [comment]
Hi. I just want to make a connection to https://github.com/serde-rs/serde/pull/2580, because I have been blocked from serde organization for reacting to the PR.

I do not want to pollute this discussion further. Please follow this issue https://github.com/andreisilviudragnea/serde-blocked/issues/2 for details.

**Comment by sagudev** [comment]
I voted 👎🏾, because there are workarounds that this change would disable: https://github.com/serde-rs/serde/issues/2538#issuecomment-1684946356

Also taking options away isn't any better and we should wait for things to cold down.

**Reply by pinkforest** [reply]
That work/around is DoA in any case regardless what happens here.

**Comment by mati865** [comment]
How about declaring serde 1.0.171 as the latest "officially" supported and testing it on CI (by running `cargo update` to pin specific version) but not limiting the users who accept bundled binary?
I'll surely restrict its version in binaries because that's the right place to do that but I can understand users who do not care much about it.

**Comment by uklotzde** [comment]
I hope that everyone voting with *thumbs up* is aware of the implications of their decision? Effectively causing a split of the whole Rust ecosystem that has dependencies on serde. This is inevitable if only some but not all crates decide to restrict the versions of serde. You could only use the libraries from either side, but not both. More and more version conflicts will arise.

Personally I don't understand why libraries try to act as a guardian for their users without any necessity. Upper version bounds could be applied by binaries in their local scope without affecting everyone else.

**Reply by lu-zero** [reply]
For many practical purposes serde with higher version number is broken on linux-x86_64, so pinning the last working one while sorting out what can be done to unbreak it is sensible.

If we can come up with a solution to make everybody happy the next release the pinning can be relaxed. 

I'd suggest to make cargo directly interact with sccache so not just syn can be cached but also the proc-macros using it could. But right now I do not have time to play with the concept :(

**Reply by AlexTMjugador** [reply]
Let me suggest an alternative view of the implications of "libraries acting as a guardian for their users", which I think is an interesting point of view.

This time, `serde_derive` tried to "act as a guardian" of build performance improvements, thinking that its users would get over it, or agree with it. It can be argued that decision was a bad judgement call.

What prevents `time` from making similar misjudgements in the future if it insists on "acting as a guardian" over issues that are not its responsibility, nor does it have the full knowledge to decide properly? If anything, this `serde_derive` situation makes me wary of how a developer can single-handedly [make it into news articles for bad reasons](https://www.bleepingcomputer.com/news/security/rust-devs-push-back-as-serde-project-ships-precompiled-binaries/).

I reiterate my desire to fix problems as broadly, correctly, and as close to the source as possible. Ill-conceived workarounds are, in my opinion, a recipe for burnout, churn, and inefficiency.

**Comment by pacak** [comment]
There's a chance that the situation will be resolved: https://github.com/serde-rs/serde/issues/2584#issuecomment-1685416231

**Comment by pinkforest** [comment]
Ok the issue has been resolved - see:
https://github.com/time-rs/time/issues/614
https://github.com/time-rs/time/pull/615

We've spent some time auditing the previous versions.

**Reply by Be-ing** [reply]
I speculate that the strong social pressure from @jhpratt's actions with this crate contributed to this outcome.

**Reply by AlexTMjugador** [reply]
> I speculate that the strong social pressure from @jhpratt's actions with this crate contributed to this outcome.

I agree. On a personal note, I'm thankful to Jacob for bringing more eyes to the situation (and even making me think more deeply about security). But I would like to meet him for other reasons, not because things are breaking and bad workarounds are being made 😅

**Comment by uklotzde** [comment]
> I speculate that the strong social pressure from @jhpratt's actions with this crate contributed to this outcome.

That's not an excuse for publishing a patch release without review that breaks builds.

**Reply by ssokolow** [reply]
1. Serde doesn't follow semver. It'll likely be on 1.0.x forever now that it's hit major version 1 because it merges minor and patch bumps into a single number and only breaks compatibility by accident.
2. dtolnay probably didn't think it *was* a breaking change for anyone.

**Comment by jhpratt** [comment]
As the deadline how now been reached, the vote is over. The result is clear: 121 in favor (70%) and 51 opposed (30%). I have saved this page with archive.org for verifiability: [link](https://web.archive.org/web/20230822000055/https://github.com/time-rs/time/discussions/611).

Given this result, the versions of `serde` that contain unreproduced executables (via `serde_derive`) will remain unsupported. As there was a release last night that removed this, expect the most recent versions of `serde` to be supported in the near future. Unfortunately, Cargo does not support multiple ranges, so the older versions of `serde` will be unsupported. However, if and when the binaries are fully reproduced, the range will be expanded to what it previously was.

Thank you to everyone who provided input! There were _far_ more people who voted than I expected. I will be closing this discussion as the poll has ended, but it will not be locked (such that further comments can be made).
